src/mlang/m_ir/mir.ml

```diff
@@ -411,6 +411,16 @@ let find_var_by_name (p : program) (name : string Pos.marked) : Variable.t =
            (Pos.VarNameToID.find name p.program_idmap))
     with Not_found -> Errors.raise_spanned_error "unknown variable" (Pos.get_position name))
 
+let find_vars_by_io (p : program) (io_to_find : io) : VariableDict.t =
+  let rules_as_list = RuleMap.bindings p.program_rules in
+  let rule_vars =
+    List.concat_map (fun (_,rd) -> rd.rule_vars) rules_as_list in
+  let filtered_list =
+    List.filter_map  (fun (var_id, var_data) -> if var_data.var_io = io_to_find then Some var_id else None) rule_vars in
+  VariableDict.filter
+    (fun vid _ -> List.exists (fun list_vid -> list_vid = vid) filtered_list)
+    p.program_vars
```

Comment:
Refactor this function to use maps, so as to have line 421 Map.mem rather than List.exists